File: mobileChatExamples/androidChatExample/app/src/main/java/com/amazon/connect/chat/androidchatexample/utils/CommonUtils.kt
package com.amazon.connect.chat.androidchatexample.utils

import com.amazon.connect.chat.sdk.model.ContentType
import com.amazon.connect.chat.sdk.model.Event
import com.amazon.connect.chat.sdk.model.Message
import com.amazon.connect.chat.sdk.model.MessageDirection
import com.amazon.connect.chat.sdk.model.MessageStatus
import com.amazon.connect.chat.sdk.model.TranscriptItem
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object CommonUtils {

    fun formatTime(timeStamp: String): String {
        val utcFormatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

        val date = utcFormatter.parse(timeStamp)
        return if (date != null) {
            val localFormatter = SimpleDateFormat("HH:mm", Locale.getDefault()).apply {
                timeZone = TimeZone.getDefault()
            }
            localFormatter.format(date)
        } else {
            timeStamp
        }
    }

    fun formatDate(currentTimeMillis: Long, forLogs: Boolean = false): String {
        val date = Date(currentTimeMillis)
        var utcFormatter: SimpleDateFormat? = null
        if (forLogs) {
            utcFormatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
        } else {
            utcFormatter = SimpleDateFormat("yyyy-MM-dd_HH-mm", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
        }
        return utcFormatter.format(date)
    }

    fun getMessageDirection(transcriptItem: TranscriptItem) {
        when (transcriptItem) {
            is Message -> {
                val participant = transcriptItem.participant?.lowercase()
                val direction = when (participant) {
                    "customer" -> MessageDirection.OUTGOING
                    "agent", "system" -> MessageDirection.INCOMING
                    else -> MessageDirection.COMMON
                }
                transcriptItem.messageDirection = direction
            }
            is Event -> {
                val participant = transcriptItem.participant?.lowercase() ?: return
                val direction = if (transcriptItem.contentType == ContentType.TYPING.type) {
                    when (participant) {
                        "customer" -> MessageDirection.OUTGOING
                        else -> MessageDirection.INCOMING
                    }
                } else {
                    MessageDirection.COMMON
                }
                transcriptItem.eventDirection = direction
            }
        }
    }

    fun customizeEvent(event: Event) {
        val displayNameOrParticipant = if (!event.displayName.isNullOrEmpty()) {
            event.displayName
        } else {
            event.participant
        } ?: "SYSTEM"

        when (event.contentType) {
            ContentType.JOINED.type -> {
                event.text = "$displayNameOrParticipant has joined the chat"
                event.participant = "System"
            }
            ContentType.LEFT.type -> {
                event.text = "$displayNameOrParticipant has left the chat"
                event.participant = "System"
            }
            ContentType.ENDED.type -> {
                event.text = "The chat has ended"
                event.participant = "System"
            }
            else -> {
                // No customization needed for other content types
            }
        }
    }


    fun customMessageStatus(status: MessageStatus?): String {
        return when (status) {
            MessageStatus.Delivered -> "Delivered"
            MessageStatus.Read -> "Read"
            MessageStatus.Sending -> "Sending"
            MessageStatus.Failed -> "Failed to send"
            MessageStatus.Sent -> "Sent"
            MessageStatus.Custom -> status.customValue ?: "Custom status"
            else -> ""  // Returning empty string for unknown or null status
        }
    }
}

File: mobileChatExamples/iOSChatExample/AmazonConnectChatIOSDemo/Utils/CommonUtils.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0


import Foundation
import AmazonConnectChatIOS

struct CommonUtils {

    static func formatTime(_ timeStamp: String) -> String? {
        let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
            dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
            dateFormatter.locale = Locale(identifier: "en_US_POSIX") // Use a POSIX locale for fixed format parsing
            
            // Attempt to convert UTC time to Date
            guard let utcDate = dateFormatter.date(from: timeStamp) else {
                return timeStamp // Return the original timestamp as a fallback
            }
            
            // Set the date formatter's timezone to the device's local timezone and format for display
            dateFormatter.timeZone = TimeZone.current
            dateFormatter.dateFormat = "HH:mm"
            
            // Convert to local time or fallback to UTC time string
            let localTime = dateFormatter.string(from: utcDate)
            return localTime.isEmpty ? timeStamp : localTime
    }
    
    
    static func getMessageDirection(from transcriptItem: TranscriptItem) {
        if let message = transcriptItem as? Message {
            let participant = message.participant
            let direction: MessageDirection
            switch participant.lowercased() {
            case "customer":
                direction = .Outgoing
            case "agent":
                direction = .Incoming
            case "system":
                direction = .Incoming
            default:
                direction = .Common
            }
            message.messageDirection = direction
        } else if let event = transcriptItem as? Event {
            guard let participant = event.participant else { return }
            var direction: MessageDirection
            direction = .Common
            if event.contentType == ContentType.typing.rawValue {
                switch participant.lowercased() {
                case "customer":
                    direction = .Outgoing
                default:
                    direction = .Incoming
                }
            }
            event.eventDirection = direction
        }
    }
    
    static func customizeEvent(from event: Event){
        let displayNameOrParticipant = (event.displayName?.isEmpty == false ? event.displayName : event.participant) ?? "SYSTEM"
        switch event.contentType {
        case ContentType.joined.rawValue:
            event.text = "\(displayNameOrParticipant) has joined the chat"
            event.participant = "System"
        case ContentType.left.rawValue:
            event.text = "\(displayNameOrParticipant) has left the chat"
            event.participant = "System"
        case ContentType.ended.rawValue:
            event.text = "The chat has ended"
            event.participant = "System"
        case ContentType.participantIdle.rawValue:
            event.text = "\(displayNameOrParticipant) has gone idle"
            event.participant = "System"
        case ContentType.participantReturned.rawValue:
            event.text = "\(displayNameOrParticipant) has returned"
            event.participant = "System"
        case ContentType.autoDisconnection.rawValue:
            event.text = "\(displayNameOrParticipant) was automatically disconnected"
            event.participant = "System"
        default:
            break
        }
    }
    
    static func customMessageStatus(for status: MessageStatus?) -> String {
        switch status {
        case .Delivered:
            return "Delivered"
        case .Read:
            return "Read"
        case .Sending:
            return "Sending"
        case .Failed:
            return "Failed to send"
        case .Sent:
            return "Sent"
        default :
            return ""  // Leaving it empty for unknown status
        }
    }
}
